classes/invitation.php

```diff
@@ -0,0 +1,210 @@
+     */
+    public function __construct($invitation) {
+        $this->invitation = $invitation;
+    }
+
+    /**
+     * Get the display string to show on the module page.
+     *
+     * @param int $coursemoduleid Course module where the user will view the invitation.
+     * @param int|null $userid Optionally supply the intended user to view the string. Defaults to global $USER.
+     * @return string|null
+     */
+    public function get_display_string(int $coursemoduleid, int $userid = null) {
+        if (empty($this->invitation)) {
+            return null;
+        }
+        // If regex patterns are disabled, return the raw zoom meeting invitation.
+        if (!get_config('zoom', 'invitationregexenabled')) {
+            return $this->invitation;
+        }
+        $displaystring = $this->invitation;
+        try {
+            // If setting enabled, strip the invite message.
+            if (get_config('zoom', 'invitationremoveinvite')) {
+                $displaystring = $this->remove_element($displaystring, 'invite');
+            }
+            // Check user capabilities, and remove parts of the invitation they don't have permission to view.
+            if (!has_capability('mod/zoom:viewjoinurl', \context_module::instance($coursemoduleid), $userid)) {
+                $displaystring = $this->remove_element($displaystring, 'joinurl');
+            }
+            if (!has_capability('mod/zoom:viewdialin', \context_module::instance($coursemoduleid), $userid)) {
+                $displaystring = $this->remove_element($displaystring, 'onetapmobile');
+                $displaystring = $this->remove_element($displaystring, 'dialin');
+                $displaystring = $this->remove_element($displaystring, 'sip');
+                $displaystring = $this->remove_element($displaystring, 'h323');
+            } else {
+                // Fix the formatting of the onetapmobile section if it exists.
+                $displaystring = $this->add_paragraph_break_above_element($displaystring, 'onetapmobile');
+            }
+        } catch (\moodle_exception $e) {
+            // If the regex parsing fails, log a debugging message and return the whole invitation.
+            debugging($e->getMessage(), DEBUG_DEVELOPER);
+            return $this->invitation;
+        }
+        $displaystring = trim($this->clean_paragraphs($displaystring));
+        return $displaystring;
+    }
+
+    /**
+     * Remove instances of a zoom invitation element using a regex pattern.
+     *
+     * @param string $invitation
+     * @param string $element
+     * @return string
+     *
+     * @throws \coding_exception
+     * @throws \dml_exception
+     * @throws \moodle_exception
+     */
+    private function remove_element(string $invitation, string $element): string {
+        global $PAGE;
+        $configregex = $this->get_config_invitation_regex();
+        if (!in_array($element, array_keys($configregex))) {
```

Comment:
Option 1: If you want to make sure they have specified a non-empty string before we try to remove the element.
```suggestion
        if (empty($configregex[$element])) {
```
Option 2: If for some reason you want to allow empty string `''` and the string zero `'0'`.
```suggestion
        if (!isset($configregex[$element])) {
```

Reply:
Hmm... Option 1 is very similar to the code in add_paragraph_break_above_element(), so is there a reason that you did not use Option 1?

Reply:
This is already handled by `preg_replace` (line 112-114). It will return 'null' if the regex is invalid which includes an empty string.

Reply:
I guess I'd rather not invoke preg_replace if it's guaranteed to fail. The functions `in_array` and `array_keys` are a longer route to answer a simple question and are less legible about the purpose than the language construct `empty`.

Reply:
I can add the empty check as well if you insist. I would point out that empty only checks for one way that preg_replace could fail so it would not remove or replace how the preg_replace is currently handled.

Reply:
I feel like there is a miscommunication here. I'm sorry for upsetting you - I'm not trying to force my way here.

I'm not suggesting removing/replacing/changing preg_replace and I'm not suggesting adding another check. I'm specifically suggesting to use `empty($configregex[$element])` in place of `!in_array($element, array_keys($configregex))`. My suggestion matches code that you already wrote in a different function, so I asked if there was a reason to use the current more-complex check to achieve the same goal.

Reply:
I'm sorry if I came across as upset. I'll try to explain my intention better.

The main reason why consciously chose to only check the keys, and not use `empty()` was I wanted to give as much information back to the user via the debugging commands as getting the regex patterns right can be a bit finicky. 

Doing a simple replacement of checking array_keys with using `empty()` would actually make the error messages not make sense, and it would need to be refactored. If a Moodle user saved an empty string as the regex pattern for example, and we are using the `empty()` command, they would get a coding exception pointing them to look at the hardcoded element options. This is misleading and not the reason the exception should have been thrown.

The coding exception is only meant to be thrown if the element string provided does not match one of the element string in the default array, which means that discreetly checking the keys is the best way to confirm this.

If we also wanted to do a check to validate that the pattern is not empty (not a coding error), we could add an additional if statement to then call another debugging message. I personally think this would be unnecessary duplication of code, and my personal preference is to not duplicate verification of input when it is already handled natively by the language. If we were to add verification that the pattern is not empty, should we also add verification that the string is a valid regex pattern as well?

Reply:
I'll also add that `preg_replace` and `preg_match` do not have the same behaviour for handling malformed regex which is why the pattern and the output of the preg_ methods are not checked in the same way.

Additionally, this is the function that is potentially called for every element so should be responsible for providing information back to the user about any errors.

The `add_paragraph_break_above_element` is only called by one element at the moment and would possibly confuse users if it gave as much verbose information as `remove_element` (which would most likely be duplicated).

Reply:
I can change it to use your suggestion `isset()` as it would still allow the regex pattern check to be handled by `preg_replace`. I suppose I considered that checking the array keys made it more obvious of what was actually being checked, but it seems I may have been wrong in this regard.

Reply:
Thank you for explaining, I understand this much better now.

It's not that we're trying to guard against a missing regex, it's that we want to notify a programmer in the future who tries to call remove_element using an element key that isn't defined by our default list of element-keyed regexes.

In fact, when the regex is missing, you're saying that we specifically *want* it to fail using the invitationmodificationfailed message instead and we're relying on the preg_replace function call to do all of that.

These details weren't immediately obvious, but I understand them now.

With all that in mind, I think that Option 2 is viable as long as the configuration values exist. If we're expecting the configuration values might not exist and this configregex might have null values, Option 3 would be:
```suggestion
        if (!array_key_exists($element, $configregex)) {
```

Reply:
I think I like option 3 better as it is safest. Should have used it in the first place. I've pushed the change up. Thanks for your help.